ui/app/src/components/DatasourceList/DatasourceDrawer.tsx

```diff
@@ -0,0 +1,97 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { Datasource } from '@perses-dev/core';
+import { DispatchWithoutAction, useState } from 'react';
+import { DiscardChangesConfirmationDialog, Drawer, ErrorAlert, ErrorBoundary } from '@perses-dev/components';
+import { PluginRegistry } from '@perses-dev/plugin-system';
+import { bundledPluginLoader } from '../../model/bundled-plugins';
+import { DeleteDatasourceDialog } from '../dialogs/DeleteDatasourceDialog';
+import { DatasourceEditorForm } from './DatasourceEditorForm';
+
+interface DatasourceDrawerProps {
+  datasource: Datasource;
+  isOpen: boolean;
+  saveAction: string;
+  onSave: (datasource: Datasource) => void;
+  onClose: DispatchWithoutAction;
+}
+
+export function DatasourceDrawer(props: DatasourceDrawerProps) {
+  const { datasource, isOpen, saveAction, onSave, onClose } = props;
+  const [isDeleteDatasourceDialogStateOpened, setDeleteDatasourceDialogStateOpened] = useState<boolean>(false);
+  const [isDiscardDialogStateOpened, setDiscardDialogStateOpened] = useState<boolean>(false);
+  const [hasDraftChanges, setHasDraftChanges] = useState<boolean>(false);
+
+  const handleSave = (datasource: Datasource) => {
+    setHasDraftChanges(false);
+    onSave(datasource);
+  };
+
+  const handleClose = () => {
+    if (hasDraftChanges) {
+      setDiscardDialogStateOpened(true);
+    } else {
+      onClose();
+    }
+  };
+
+  const handleDelete = () => {
+    setHasDraftChanges(false);
+    onClose();
+  };
+
+  const handleDiscard = () => {
+    setHasDraftChanges(false);
+    setDiscardDialogStateOpened(false);
+    onClose();
+  };
+
+  return (
+    <Drawer isOpen={isOpen} onClose={handleClose} data-testid="datasource-editor">
+      <ErrorBoundary FallbackComponent={ErrorAlert}>
+        <PluginRegistry
+          pluginLoader={bundledPluginLoader}
+          // TODO this required field is useless here
+          defaultPluginKinds={{
+            Panel: 'TimeSeriesChart',
+            TimeSeriesQuery: 'PrometheusTimeSeriesQuery',
+          }}
+        >
+          {isOpen && (
+            <DatasourceEditorForm
+              initialDatasource={datasource}
+              saveAction={saveAction}
+              flagAsDraft={() => setHasDraftChanges(true)}
+              onSave={handleSave}
+              onCancel={handleClose}
+              onDelete={saveAction == 'Update' ? () => setDeleteDatasourceDialogStateOpened(true) : undefined}
+            />
+          )}
+        </PluginRegistry>
+        <DeleteDatasourceDialog
+          open={isDeleteDatasourceDialogStateOpened}
+          onClose={() => setDeleteDatasourceDialogStateOpened(false)}
+          onDelete={handleDelete}
+          datasource={datasource}
+        />
+        <DiscardChangesConfirmationDialog
```

Comment:
If you check on that PR https://github.com/perses/perses/pull/1290, I moved the discard change behavior into the VariableEditorForm. This can remove some duplication if you use it somewhere else and prevent you from using this flagAsDraft over there no ?

Reply:
I see that you went for a comparison based on JSON.stringify to detect if there was changes.. Yes I think it's better than the solution I did, I'll reuse this thanks

Reply:
Not perfect maybe but that's the trick I found for deep compares. Maybe there would still be an issue if you have inner map ? I'm not sure if JSON.stringify orders maps by key by default or if it's kinda random here